File: cpp/recursion_bruteforce/recursion_bruteforce/subset.cpp
#include "subset.h"

#include <iostream>
using namespace std;

const int N = 4;
static int b[N];

void subset(int k)
{
	// cout << "k : " << k << "\t : " << b[0] << b[1] << b[2] << b[3] << endl;
	if (k < N)
	{
		for (int i = 0; i < 2; i++)
		{
			b[k] = i; // 0 or 1
			subset(k + 1);
		}
	}
	else
	{
		int initial = 1;
		cout << "{ ";
		for (int i = 0; i < N; i++)
		{
			if (b[i])
			{
				if (initial)
					initial = 0;
				else
					cout << ", ";
				cout << i;
			}
		}
		cout << " }" <<  endl;
	}
}

File: programming-contest/subset/subset.py
def subset(k, N, a):
    if k < N:
        # 全ての選択肢を再帰で取得する
        a[k] = 0
        subset(k + 1, N, a)
        a[k] = 1
        subset(k + 1, N, a) 
    else:
        # どの要素が選択されているか？
        for i in range(4):
            print(a[i], end=' ')
        print()

        # 表示処理
        print('{', end='')
        initial = 1
        for i in range(N):
            if a[i]:
                if initial != 0:
                    initial = 0
                else:
                    print(', ', end='')
                print(i, end='')
        print("}")


if __name__ == "__main__":
    subset(0, 4, [0]*4)